Completed updating code for latest rasa v2.4

kmanadkat opened the pull request:
# Update supporting Rasa v2.4.x

This update includes mostly Form & Custom action related changes. To make it rasa v2.4.x friendly completely, I went ahead and changed files from markdown to yaml.
```
stories.md -> stories.yml
nlu.md -> nlu.yml
++ rules.yml

__init__.py -> data/__init__.py
actions.py -> data/actions.yml
```
Also made tiny update in `config.yml` updating `FormPolicy` to `RulePolicy` as per latest [docs](https://rasa.com/docs/rasa/forms#usage)

<br />

## Testing & Quality Assurance of Update

**rasa train**

<img width="1594" alt="01 rasa-train-success" src="https://user-images.githubusercontent.com/25281055/114272267-26ce5c80-9a33-11eb-87e7-95d5d87484b3.png">


**rasa run actions**

<img width="1016" alt="02 rasa-run-actions-success" src="https://user-images.githubusercontent.com/25281055/114272602-72cdd100-9a34-11eb-8e58-1ede160a17e8.png">


**rasa shell - happy flow 1 - affirmed for exercise**

<img width="1525" alt="03 rasa-shell-happy-flow" src="https://user-images.githubusercontent.com/25281055/114272610-782b1b80-9a34-11eb-9eca-676914f5f838.png">


**rasa shell - happy flow 2 - denied for exercise**

<img width="1528" alt="04 rasa-shell-happy-flow-no-exercise" src="https://user-images.githubusercontent.com/25281055/114272616-7bbea280-9a34-11eb-92e2-504b5840c13e.png">


**rasa shell - sad flow 1 - going out of scope & coming back**

<img width="1523" alt="05 rasa-shell-sad-path-1" src="https://user-images.githubusercontent.com/25281055/114272620-7f522980-9a34-11eb-964c-e79b429cf42b.png">


**rasa shell - sad flow 2 - going out of scope & denied to proceed**

<img width="1525" alt="06 rasa-shell-sad-path-2" src="https://user-images.githubusercontent.com/25281055/114272624-82e5b080-9a34-11eb-9918-c1a6756e4fa7.png">


Rasa [Documentation](https://rasa.com/docs/), examples on [github](https://github.com/RasaHQ/rasa/tree/main/examples) & amazing community [forum](https://forum.rasa.com/latest) helped me update code base.

kmanadkat committed:
Completed updating code for latest rasa v2.4

I have tested bot for happy flow & 2 sad flows. It has worked well. This update includes mostly Form & Custom action related changes.

kmanadkat commented:
hi @karen-white & @rgstephens would you please review above pull request?

sara-tagger commented:
Thanks for submitting a pull request 🚀 @joejuzl will take a look at it as soon as possible ✨

karen-white commented:
Hi @kmanadkat, thanks for taking the time and effort to make these updates. This bot is a little bit of a special case because there is a [tutorial](https://blog.rasa.com/how-to-build-your-first-rasa-form/) associated with it. When Rasa 2.0 came out, we made the decision to leave the 1.0-compatible tutorial as-is and create a [new blog post showing how to upgrade the bot](https://blog.rasa.com/updating-to-rasa-open-source-2-0/). 

I'd like to keep the main branch consistent with the tutorial, but I'd be comfortable with pushing your changes to a 2.x branch. We could add a note to the readme that calls out the 2.4 branch for anyone who wants to see the upgraded version. How does that sound?

kmanadkat commented:
@karen-white thanks for your time, yes I totally agree with you :) 

I tried creating pull requests from forked repository branch to a different branch, it is not allowing me. I am sorry this is my first open source contribution. I think if you create a new branch for update from master then I will be able to submit PR :)

<img width="1912" alt="Screenshot 2021-04-13 at 12 07 53 AM" src="https://user-images.githubusercontent.com/25281055/114444287-4e1c5980-9bec-11eb-9d28-430b1e7b70b4.png">

kmanadkat commented:
And I would be happy to submit update for Readme in both branches & 
It would be my honour to write a whole new article (for 2.4.x) helpful for beginners to understand wellness bot if you or someone from rasa team helps in reviewing my write up 🌈

karen-white commented:
Sounds good @kmanadkat ! I changed the base branch. Did you want to go ahead and update the readme before I merge?

Since we have [this blog post](https://blog.rasa.com/updating-to-rasa-open-source-2-0/) already, maybe it would make sense to do a tutorial that's about your first-person experience upgrading the bot (and making your first open source contribution 💯 )? Tagging @EmmaWightman for ideas :)

kmanadkat commented:
Readme updated and commits are added to this pull request, please review.

Would you like to have github pages demo site with a video? [check this one](https://kmanadkat.github.io/rasa-wellness-bot/)

If you prefer, I can add demonstration website in `/docs` path of `update-to-rasa-2.4/master` branch, just like above website. Then you can enable github pages in `/docs` folder from repository `pages` settings :)

> Ofc would change hyperlinks everywhere so it points to rasa official docs and github repositories.

kmanadkat committed:
Added new conversation image & demo video

kmanadkat committed:
Readme file update for rasa 2.4

Added Installation and Setup steps in detail

kmanadkat committed:
Fixed typos

kmanadkat commented on README.md:
````diff
@@ -1,18 +1,36 @@
+rasa shell
+```
+
+![wellness-bot-conversation](./images/bot_conversation_2.4.png)
````
@karen-white preview on line 36 should work after code is merged because it works [here](https://github.com/kmanadkat/wellness-check-bot/tree/update-to-rasa-2.4/master)

karen-white commented:
thanks @kmanadkat ! I will take a closer look in the next few days and review - definitely by the end of this week :)

kmanadkat committed:
Removed video demo of bot, not required right now.

We can put it back if we go with github pages

kmanadkat commented:
Hey @karen-white looking forward to hear from you this weekend :)

karen-white commented on domain.yml:
```diff
@@ -15,63 +37,95 @@ entities:
+
   utter_ask_continue:
-  - text: "Sorry, I don't quite understand. Do you want to continue?"
+  - text: Sorry, I don't quire understand. Do you want to continue?
```
Looks like there's a small typo here :)

kmanadkat replied:
Nice catch, so silly of me, fixing it  :)

karen-white commented on domain.yml:
```diff
@@ -15,63 +37,95 @@ entities:
-  - text: "I am a bot, powered by Rasa."
-  
-  utter_ask_confirm_exercise:
+  - text: I am just another bot trying to help you.
```
I see the intent and response that's been added for bot_challenge/utter_iamabot, but no rule to create the dialogue?

kmanadkat replied:
my bad. I am fixing it with below things
**domain.yml**
```yml
  utter_iamabot:
  - text: I am a bot, powered by Rasa.
```

**data/rules.yml**
```yml
- rule: Challenging Rasa Bot
  steps:
  - intent: bot_challenge
  - action: utter_iamabot
```

karen-white commented on domain.yml:
```diff
@@ -15,63 +37,95 @@ entities:
-  - text: "Hi! It's time for your daily wellness check. Tracking healthy habits is a great way to measure your progress over time. Would you like to answer a few questions about your health?"
+  - text: Hi! It's time for your daily wellness check. Tracking healthy habits is a great way to measure your progress over time. Would you like to answer few questions about your health?
+
+  utter_cheer_up:
```
I see that the moodbot dialogues have been added to the wellness check bot. What was the reason for including these?

kmanadkat replied:
Yes you are right. I am removing all mood bot utterances. I started from `rasa init` & must have forgotten to remove default things.

karen-white commented on domain.yml:
```diff
@@ -15,63 +37,95 @@ entities:
-forms:
-  - health_form
\ No newline at end of file
+  carry_over_slots_to_new_session: true
```
Curious about why you'd want to have slots carry over to new sessions? The use case is to have the user fill this form out once per day, so it seems like you'd want it to re-set after each session rather than each form-fill. I see you've also added an action_restart to each story - wondering what the reasoning is there?

kmanadkat replied:
`action_restart` is for resetting the form, in case user wants to retry filling the form then we would want bot to recollect fresh responses instead of using information from prev interaction.

I see your point for `carry_over_slots_to_new_session: true` I am making it false :)

karen-white commented on domain.yml:
```diff
@@ -15,63 +37,95 @@ entities:
+  utter_ask_health_form_diet:
+  - text: Did you stick to a healthy diet 🥦 yesterday?
+
+  utter_ask_health_form_stress:
```
Looks like the buttons in the response got dropped. Was that intended?

kmanadkat replied:
yes, I did not put buttons intentionally. But now I feel, buttons would make it easy for user to answer. Thanks for bringing this up, adding buttons back :)

kmanadkat replied:
I changed response to below
```yml
  utter_ask_health_form_stress:
  - text: Is your stress level low, medium or high 🧘🏼?
    buttons:
    - title: "Low"
      payload: '/inform{"stress": "low"}'
    - title: "Medium"
      payload: '/inform{"stress": "medium"}'
    - title: "High"
      payload: '/inform{"stress": "high"}'
```
We have `stress` defined in slots as well as entities.

I am facing an error with buttons. Python is trying to interpret `{..}` as placeholder for string interpolation. It shows error in console. Reading rasa docs, I understood that payload should be a Json like this -> `{"entity": "value"}`

<img width="1513" alt="Screenshot 2021-04-20 at 2 33 52 PM" src="https://user-images.githubusercontent.com/25281055/115369192-7e17be00-a1e5-11eb-8ba1-341752c31640.png">

Though, if we ignore this error, bot seems to work. What do you suggest here?

kmanadkat replied:
It works without throwing error. I referred [this stackoverflow answer](https://stackoverflow.com/a/5466478/7475099).

```yml
  utter_ask_health_form_stress:
  - text: Is your stress level low, medium or high 🧘🏼?
    buttons:
    - title: "Low"
      payload: '/inform{{"stress": "low"}}'
    - title: "Medium"
      payload: '/inform{{"stress": "medium"}}'
    - title: "High"
      payload: '/inform{{"stress": "high"}}'
```

(added `{{...}}` double curly brackets.)

<img width="1434" alt="Screenshot 2021-04-20 at 2 55 55 PM" src="https://user-images.githubusercontent.com/25281055/115372444-84f40000-a1e8-11eb-9516-45d448998c63.png">

karen-white reviewed (changes requested):
Looks good! Left a few questions :)

kmanadkat committed:
Resolutions for PR review

- quite typo fix
- added rule for bot challenge
- removed mood bot dialogues
- carry slots value to next session set to false
- made buttons in stress level utterance
- added new screenshot of rasa shell

kmanadkat commented:
Hi @karen-white that was an amazing review. I did several silly mistakes 😅, thanks for your time. I have worked on all of them & just pushed [commit](https://github.com/RasaHQ/wellness-check-bot/pull/3/commits/5b1182361112dbdcca58fcacb6c506ddca73d229) 
```yml
- quite typo fix
- added rule for bot challenge
- removed mood bot dialogues
- carry slots value to next session set to false
- made buttons in stress level utterance
- added new screenshot of rasa shell
```

Please review it, and we can go next with anything you say.

karen-white commented:
@kmanadkat Looks great! Thanks for making those changes (and for helping us keep this tutorial bot up to date) :)

kmanadkat commented:
@karen-white that was amazing to see PR merged! I am eager to do more contribution with such welcoming community like of rasa :)
Open to any ideas you have.
